src/core/model/reducer.ts

```diff
@@ -92,11 +75,21 @@ export const uiSchemaReducer = (
 
 export const combinedReducer = (state: EditorState, action: CombinedAction) => {
   switch (action.type) {
+    case SET_SCHEMA:
+      return buildAndLinkUISchema(
+        buildSchemaTree(action.schema),
+        state.uiSchema
+      );
+    case SET_UISCHEMA:
+      return buildAndLinkUISchema(
+        cleanLinkedElements(state.schema),
```

Comment:
We double-clone the schema here (once in `cleanLinkedElements`) and once in `buildAndLinkUiSchema`. In the other actions we construct new schemas and immediately clone them again. 

It's also not optimal that we hand in both regular UI schemas and already linked UI schemas to the same method. This also leads to a weird consequence: All ui schema uuids will be reset every time the JSON schema is manually adapted.

I would like to suggest to split this into two functions:
 * `buildUiSchema` (or rename to `construct` or something) which takes regular ui schemas and returns a new one with parents and uuids set, i.e. `UISchemaElement` to `EditorUISchemaElement` conversion`.
 * `linkSchemas` which takes two already cloned (i.e. save to edit) schemas and makes sure that all links are set correctly once the method is finished.

These can then be used in the appropriate places.